pkg/lib/feature/forgotpassword/provider.go

```diff
@@ -247,21 +247,18 @@ func (p *Provider) sendSMS(phone string, code string) (err error) {
 	return
 }
 
-// ResetPassword consumes code and reset password to newPassword.
+// CheckResetPasswordCode check whether the code is valid.
 // If the code is invalid, ErrInvalidCode is returned.
 // If the code is found but expired, ErrExpiredCode is returned.
 // if the code is found but used, ErrUsedCode is returned.
-// Otherwise, the password is reset to newPassword.
-// newPassword is checked against the password policy so
-// password policy error may also be returned.
-func (p *Provider) ResetPasswordByCode(codeStr string, newPassword string) (err error) {
+func (p *Provider) checkResetPasswordCode(codeStr string) (code *Code, err error) {
 	err = p.RateLimiter.TakeToken(AntiBruteForceVerifyBucket(string(p.RemoteIP)))
```

Comment:
This TakeToken step should only be done for failed verify attempt. So merely checking a correct code does not take a token. And this implies keep refreshing a page with correct code will not encounter rate limit error.

Reply:
In that case, the rate limit would be useless, since potential attacker may check correctness of code by GETting the page instead of POSTing the page.

Reply:
We count failed attempts so the attacker can only try a few times

Reply:
btw there are other places having similar issue. https://github.com/authgear/authgear-server/issues/2869 If you have no tasks on hand, please take it.